### PR TITLE
Fixes #90125 - Add a task instancePolicy to task runOptions

### LINKED ISSUE #90125: Add a task instancePolicy to task runOptions.
When a task has reached the max number of instances it can have (default is 1, can be changed with `instanceLimit`) the `instancePolicy` determines what happens when another instance is started.

History: https://github.com/microsoft/vscode/issues/32264
PR for `insanceLimit` which is relevant to anyone interested in implementing `instancePolicy`: https://github.com/microsoft/vscode/pull/89872

### PR DESCRIPTION
**Pull request by alpalla**
# Summary
Fixes #90125. Added an `instancePolicy` to `runOptions`. The `instancePolicy` is applied when creating a new instance of a task once the `instanceLimit` has been reached.
The `instancePolicy` is as follows:
1. `terminateNewest`: the newest instance is terminated
2. `terminateOldest`: the oldest instance is terminated
3. `prompt`: the QuickPick is shown and the user picks an instance to restart (this is the default option)
4. `warn`: do nothing (do not terminate an existing instance or create a new one) and show a warning that indicates that the `instanceLimit` has been reached
5. `silent`: do nothing

# Implementation notes
* any task selected to be "terminated" by the `instancePolicy` is effectively *restarted*
* the `TerminalTaskSystem` `getLastInstance()` method has been expanded in scope and is now called `getNthInstance`. As the name suggests it returns the Nth instance of a task given a number. If no index is given, it returns the last instance by default. I used a `for` loop instead of doing a `forEach` on `Object.keys` so that I could exit early once the right instance was found
* when an instance is created, slightly changed the way I previously set its `id` so as to avoid cases in which an instance of an instance was being created and the `id` would be appended with `|<COUNTER>` multiple times

# Testing changes
Create a `tasks.json` and set the `instancLimit` and `instancePolicy` properties. Here is an example configuration:
```json
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "echo",
            "type": "shell",
            "command": "sleep 20 && echo Hello",
            "runOptions": {
                "instanceLimit": 2,
                "instancePolicy": "terminateOldest"
            }
        }
    ]
}
```
Ensure that once the `instanceLimit` is reached, the chosen `instancePolicy` behaves as described at the beginning of this PR.

### P.S.
I would suggest not having the `prompt` option as the default option simply because in the default case where `instanceLimit == 1`, the user would be shown the QuickPick which would contain only one entry since there is only one active instance. Personally, I think it would make more sense to have another `instancePolicy` option which would ask the user to terminate or restart the last instance. This would maintain the current behavior and use the current terminate/restart notification, while I think being the least confusing option in the case where `runOptions` is set to all default settings.

### CODE REVIEW

**Comment by alpalla** [comment]
@alexr00 any updates regarding anything I can do to move forward with this PR? 😃

**Comment by sky126** [comment]
need this feature

File: src/vs/workbench/contrib/tasks/common/jsonSchema_v2.ts [hunk 1]
```diff
@@ -334,6 +334,12 @@ const runOptions: IJSONSchema = {
 			description: nls.localize('JsonSchema.tasks.instanceLimit', 'The number of instances of the task that are allowed to run simultaneously.'),
 			default: 1
 		},
+		instancePolicy: {
+			type: 'string',
+			enum: ['terminateNewest', 'terminateOldest', 'prompt', 'warn', 'silent'],
```
Comment (alexr00) [inline src/vs/workbench/contrib/tasks/common/jsonSchema_v2.ts:339]: Can you add an enum description for this?

**Comment by alexr00** [comment]
I haven't had a chance to do a full review yet, but I hope to next week. Can you resolve the conflicts?

**Comment by alpalla** [comment]
Sure! I'll resolve the conflicts and add that enum description :+1:

**Comment by alpalla** [comment]
@alexr00 I can't resolve the conflicts and test the resulting code locally because it seems like a couple of things are broken in the latest build. The main issues are:
1. the `instanceLimit` doesn't seem to be respected anymore (meaning I can create `K` instances of a task with a limit of `N` where `K > N`).
2. If I create n instances of a task and let all instances run to completion, it looks like they are not automatically terminated because all instances are still visible as options to terminate.

I can resolve the conflicts via GitHub without testing the resulting code first but I'm a little hesitant to do so. Let me know how I should proceed, thanks!

**Comment by onelivesleft** [comment]
Is this feature dead in the water?  Would be a great addition.

**Comment by alexr00** [comment]
Closing in favor of the newer one: https://github.com/microsoft/vscode/pull/117129

**Pull request closed without merging**